src/gc.c

```diff
@@ -0,0 +1,225 @@
+
+typedef struct GarbageCollectorCtx {
+
+  // current frequency
+  float hz;
+
+  // inverted index key name for reopening the index
+  const RedisModuleString *keyName;
+
+  // periodic timer
+  struct RMUtilTimer *timer;
+
+  GCStats stats;
+
+} GarbageCollectorCtx;
+
+GarbageCollectorCtx *NewGarbageCollector(const RedisModuleString *k, float initialHZ) {
+  GarbageCollectorCtx *gc = malloc(sizeof(*gc));
+
+  *gc = (GarbageCollectorCtx){
+      .timer = NULL, .hz = initialHZ, .keyName = k, .stats = {},
+  };
+  return gc;
+}
+
+int isRdbLoading(RedisModuleCtx *ctx) {
+  long long isLoading = 0;
+  RMUtilInfo *info = RMUtil_GetRedisInfo(ctx);
+  if (!info) {
+    return 0;
+  }
+
+  if (!RMUtilInfo_GetInt(info, "loading", &isLoading)) {
+    isLoading = 0;
+  }
+
+  RMUtilRedisInfo_Free(info);
+  return isLoading == 1;
+}
+
+void GC_PeriodicCallback(RedisModuleCtx *ctx, void *privdata) {
+
+  RedisModuleKey *idxKey = NULL;
+  RedisSearchCtx *sctx = NULL;
+  RedisModule_AutoMemory(ctx);
+  RedisModule_ThreadSafeContextLock(ctx);
+  if (isRdbLoading(ctx)) {
+    RedisModule_Log(ctx, "notice", "RDB Loading in progress, not performing GC");
+    goto end;
+  }
+  GarbageCollectorCtx *gc = privdata;
+  assert(gc);
+  sctx = NewSearchCtx(ctx, (RedisModuleString *)gc->keyName);
+  if (!sctx) {
+    RedisModule_Log(ctx, "warning", "No index spec for GC %s",
+                    RedisModule_StringPtrLen(gc->keyName, NULL));
+    goto end;
+  }
+  TimeSample ts;
+  char *term = IndexSpec_GetRandomTerm(sctx->spec, 20);
+
+  if (!term) {
+    // printf("No term for GC to inspect\n");
+    goto end;
+  }
+  RedisModule_Log(ctx, "debug", "Garbage collecting for term '%s'", term);
+  InvertedIndex *idx = Redis_OpenInvertedIndexEx(sctx, term, strlen(term), 1, &idxKey);
+  size_t totalRemoved = 0;
+  size_t totalCollected = 0;
+  if (idx) {
+    // printf("Garbage collecting for term %s\n", term);
+    int blockNum = 0;
+    int num = 100;
+
+    do {
+      size_t bytesCollected = 0;
+      size_t recordsRemoved = 0;
+
+      TimeSampler_Start(&ts);
+
+      blockNum = InvertedIndex_Repair(idx, &sctx->spec->docs, blockNum, num, &bytesCollected,
+                                      &recordsRemoved);
+      TimeSampler_End(&ts);
+      RedisModule_Log(ctx, "debug", "Repair took %lldns", TimeSampler_DurationNS(&ts));
+
+      /// update the statistics with the the number of records deleted
+      totalRemoved += recordsRemoved;
+      sctx->spec->stats.numRecords -= recordsRemoved;
+      sctx->spec->stats.invertedSize -= bytesCollected;
+      gc->stats.totalCollected += bytesCollected;
+      totalCollected += bytesCollected;
+
+      // blockNum 0 means error or we've finished
+      if (!blockNum) break;
+
+      // After each iteration we yield execution
+      // First we close the relevant keys we're touching
+      RedisModule_CloseKey(sctx->key);
+      RedisModule_CloseKey(idxKey);
+      SearchCtx_Free(sctx);
+
+      // now release the global lock
```

Comment:
Don't we have a YIELD macro for this?
On second thought, the whole reopening logic might be better done in a function.